subsys/mpsl/mpsl_init.c

```diff
@@ -127,6 +132,260 @@ static uint8_t m_config_clock_source_get(void)
+	 * given above.
+	 */
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(gpio0), okay)
+	if (strcmp(gpio_lbl, DT_LABEL(DT_NODELABEL(gpio0))) == 0) {
+		return;
+	}
+#endif
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(gpio1), okay)
+	if (strcmp(gpio_lbl, DT_LABEL(DT_NODELABEL(gpio1))) == 0) {
+		*p_gpio_pin += 32;
+		return;
+	}
+#endif
+
+	__ASSERT(false, "Unknown GPIO port DT label");
+}
+
+#endif /* IS_ENABLED(CONFIG_MPSL_FEM) */
+
+#if IS_ENABLED(CONFIG_MPSL_FEM_NRF21540_GPIO)
+static int fem_nrf21540_gpio_configure(void)
+{
+	/* FEM configuration requires gpiote and ppi channels.
+	 * Currently there is no reliable common method to dynamically
+	 * allocate such channels. FEM module needs only "some" channels
+	 * to use whichever they are, but FEM needs them for exclusive use
+	 * and does not enable them immediately.
+	 *
+	 * When common api to assign gpiote and ppi channels is available
+	 * current solution based on macros coming from Kconfig should be
+	 * reworked.
+	 */
+
+#if !DT_NODE_EXISTS(DT_NODELABEL(nrf_radio_fem))
+#error Node with label 'nrf_radio_fem' not found in the devicetree.
+#endif
+
+	mpsl_fem_nrf21540_gpio_interface_config_t cfg = {
+		.fem_config = {
+			.pa_time_gap_us  =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem),
+					tx_en_settle_time_us),
+			.lna_time_gap_us =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem),
+					rx_en_settle_time_us),
+			.pdn_settle_us   =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem),
+					pdn_settle_time_us),
+			.trx_hold_us     =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem),
+					trx_hold_time_us),
+			.pa_gain_db      =
+				CONFIG_MPSL_FEM_NRF21540_TX_GAIN_DB,
+			.lna_gain_db     =
+				CONFIG_MPSL_FEM_NRF21540_RX_GAIN_DB
+		},
+		.pa_pin_config = {
+#if DT_NODE_HAS_PROP(DT_NODELABEL(nrf_radio_fem), tx_en_gpios)
+			.enable       = true,
+			.active_high  =
+				MPSL_FEM_GPIO_POLARITY_GET(tx_en_gpios),
+			.gpio_pin     =
+				DT_GPIO_PIN(DT_NODELABEL(nrf_radio_fem),
+					    tx_en_gpios),
+			.gpiote_ch_id =
+				CONFIG_MPSL_FEM_NRF21540_GPIO_GPIOTE_TX_EN
+#else
+			MPSL_FEM_DISABLED_GPIO_CONFIG_INIT
+#endif
+		},
+		.lna_pin_config = {
+#if DT_NODE_HAS_PROP(DT_NODELABEL(nrf_radio_fem), rx_en_gpios)
+			.enable       = true,
+			.active_high  =
+				MPSL_FEM_GPIO_POLARITY_GET(rx_en_gpios),
+			.gpio_pin     =
+				DT_GPIO_PIN(DT_NODELABEL(nrf_radio_fem),
+					    rx_en_gpios),
+			.gpiote_ch_id =
+				CONFIG_MPSL_FEM_NRF21540_GPIO_GPIOTE_RX_EN
+#else
+			MPSL_FEM_DISABLED_GPIO_CONFIG_INIT
+#endif
+		},
+		.pdn_pin_config = {
+#if DT_NODE_HAS_PROP(DT_NODELABEL(nrf_radio_fem), pdn_gpios)
+			.enable       = true,
+			.active_high  =
+				MPSL_FEM_GPIO_POLARITY_GET(pdn_gpios),
+			.gpio_pin     =
+				DT_GPIO_PIN(DT_NODELABEL(nrf_radio_fem),
+					    pdn_gpios),
+			.gpiote_ch_id =
+				CONFIG_MPSL_FEM_NRF21540_GPIO_GPIOTE_PDN
+#else
+			MPSL_FEM_DISABLED_GPIO_CONFIG_INIT
+#endif
+		},
+		.ppi_channels = {
+			CONFIG_MPSL_FEM_NRF21540_GPIO_PPI_CHANNEL_0,
```

Comment:
When allocating PPI channels, this must either be either be allocated run-time like done here: https://github.com/nrfconnect/sdk-nrf/blob/master/subsys/mpsl/mpsl_init.c#L137.
Or compile time, like done here: https://github.com/nrfconnect/sdk-nrf/blob/master/subsys/bluetooth/controller/hci_driver.c#L52
Otherwise the application may end up in a situation where a PPI channel is reconfigured.

Currently there is no nice way of doing it. @anangl FYI

Reply:
Discussed offline. We intend to keep current proposal temporarily, until thread-safe channel allocator for PPI and GPIOTE is available in NCS.